engine/src/main/java/org/pentaho/di/job/Job.java

```diff
@@ -0,0 +1,2367 @@
+          } catch ( Throwable e ) {
+            log.logError( Const.getStackTracker( e ) );
+            throw new KettleException( BaseMessages.getString( PKG, "Job.Log.UnexpectedError", nextEntry.toString() ),
+                e );
+          }
+          if ( log.isBasic() ) {
+            log.logBasic( BaseMessages.getString( PKG, "Job.Log.FinishedJobEntry", nextEntry.getName(), res.getResult()
+                + "" ) );
+          }
+        }
+      }
+    }
+
+    // OK, if we run in parallel, we need to wait for all the job entries to
+    // finish...
+    //
+    if ( jobEntryCopy.isLaunchingInParallel() ) {
+      for ( int i = 0; i < threads.size(); i++ ) {
+        Thread thread = threads.get( i );
+        JobEntryCopy nextEntry = threadEntries.get( i );
+
+        try {
+          thread.join();
+        } catch ( InterruptedException e ) {
+          log.logError( jobMeta.toString(), BaseMessages.getString( PKG,
+              "Job.Log.UnexpectedErrorWhileWaitingForJobEntry", nextEntry.getName() ) );
+          threadExceptions.add( new KettleException( BaseMessages.getString( PKG,
+              "Job.Log.UnexpectedErrorWhileWaitingForJobEntry", nextEntry.getName() ), e ) );
+        }
+      }
+      // if(log.isBasic()) log.logBasic(BaseMessages.getString(PKG,
+      // "Job.Log.FinishedJobEntry",startpoint.getName(),res.getResult()+""));
+    }
+
+    // Perhaps we don't have next steps??
+    // In this case, return the previous result.
+    if ( res == null ) {
+      res = prevResult;
+    }
+
+    // See if there where any errors in the parallel execution
+    //
+    if ( threadExceptions.size() > 0 ) {
+      res.setResult( false );
+      res.setNrErrors( threadExceptions.size() );
+
+      for ( KettleException e : threadExceptions ) {
+        log.logError( jobMeta.toString(), e.getMessage(), e );
+      }
+
+      // Now throw the first Exception for good measure...
+      //
+      throw threadExceptions.poll();
+    }
+
+    // In parallel execution, we aggregate all the results, simply add them to
+    // the previous result...
+    //
+    for ( Result threadResult : threadResults ) {
+      res.add( threadResult );
+    }
+
+    // If there have been errors, logically, we need to set the result to
+    // "false"...
+    //
+    if ( res.getNrErrors() > 0 ) {
+      res.setResult( false );
+    }
+
+    return res;
+  }
+
+  /**
+   * Wait until this job has finished.
+   */
+  public void waitUntilFinished() {
+    waitUntilFinished( -1L );
+  }
+
+  /**
+   * Wait until this job has finished.
+   *
+   * @param maxMiliseconds
+   *          the maximum number of ms to wait
+   */
+  public void waitUntilFinished( long maxMiliseconds ) {
+    long time = 0L;
+    while ( isAlive() && ( time < maxMiliseconds || maxMiliseconds <= 0 ) ) {
+      try {
+        Thread.sleep( 1 );
+        time += 1;
+      } catch ( InterruptedException e ) {
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

**Problem**: InterruptedException is ignored. This can delay thread shutdown and clear the thread’s interrupt status. Only code that implements a thread’s interruption policy can swallow an interruption request.

**Fix**: Rethrow the InterruptedException or reinterrupt the current thread using *Thread.currentThread().interrupt()* so that higher-level interrupt handlers can function correctly.
 If you are wrapping the InterruptedException inside a RuntimeException, call *Thread.currentThread().interrupt()* before throwing the RuntimeException.

Learn more about [interrupts](https://docs.oracle.com/javase/tutorial/essential/concurrency/interrupt.html) and [dealing with InterruptedException](https://www.ibm.com/developerworks/java/library/j-jtp05236/index.html?ca=drs-#N10187)